### PR TITLE
Bounding Volume Hierarchy accelerator

### PR DESCRIPTION
**Pull request by nfrasser**
- [x] Primitive/Shape interface refactor
- [x] Split scene primitives out of references settings
- [x] Linear BVH implementation
- [x] Web support

### CODE REVIEW

**Comment by nfrasser** [comment]
playground

## Rendering time
Native: 18.64s user 0.04s system 389% cpu 4.796 total (yay!)
WebAssembly (Safari): 25.523s
WebAssembly (Chrome): 64.631s (come on Chrome 😞)

File: www/renderer.ts [hunk 1]
```diff
@@ -74,16 +73,20 @@ async function capture(sceneFunctionBody: string) {
...
+        lasgun.capture_hunk(i, scene, root, hunk)
+        let start = hunk.as_ptr();
+        let end = start + 1024;
         self.postMessage({
             type: 'hunk',
             value: {
                 x: hunk.x,
                 y: hunk.y,
-                data: new Uint8ClampedArray(wasm.memory.buffer, hunk.as_ptr(), 1024)
+                data: new Uint8ClampedArray(mem.slice(start, end))
```
Comment (nfrasser) [inline www/renderer.ts]: Not using `mem.slice` means the entire WASM memory buffer (several megabytes) gets copied for every 16x16 hunk of pixels that goes into `Uint8ClampedArray`. The browser was using 40+ GB (!!!) of memory before garbage collection ran.

Good times.

File: src/lib.rs [hunk 2]
```diff
@@ -66,20 +75,28 @@ pub fn capture(scene: &Scene, film: &mut Film) {
...
         // this function call ends, otherwise very bad things will happen.
+        //
+        // TODO: Pls. make this less terrifying
         let sendable_scene_ptr = UnsafeThreadWrapper(scene as *const Scene);
+        let sendable_root_ptr = UnsafeThreadWrapper(unsafe {
+            // I am so, so, so sorry, I need this to get sendable_root_ptr
+            // across the thread boundary. I promise I super-quadruple-checked
+            // that this is safe.
+            std::mem::transmute::<&Accel<'_>, &Accel<'static>>(&root)
+        } as *const Accel);
```
Comment (nfrasser) [inline src/lib.rs]: I promise I'll fix this at some point 😨

**Comment by davidrusu** [comment]
@nfrasser Did you test at all on firefox? them mozilla folks seem to be leading the way on wasm performance

File: www/renderer.ts (hunk 1 above)
Comment (davidrusu) [reply www/renderer.ts]: *do not want!*

**Comment by nfrasser** [comment]
@davidrusu I haven't for the bunny but Firefox has been comparable to Safari/Webkit in other scenes I've tried

**Comment by nfrasser** [comment]
@davidrusu ran Firefox just for you, it took 30.016s

**Comment by davidrusu** [comment]
Much improved over the last benchmark, great work :)

**Pull request merged**